src/cache-manager.js

```diff
@@ -50,22 +50,39 @@ class CacheManager {
     assert(rawUrl);
     this.debug(`putting ${rawUrl}`);
 
+    let mutex = new Mutex(this.redis, this.lockKey(rawUrl));
```

Comment:
we basically already have locking through the `pending` status in redis.  If we didn't, then cloud-mirror would be useless.  Say we had a 100mb artifact that we needed to transfer.  If we didn't do any form of locking, and we got 10 requests for that file, we'd strart 10 different 100mb transfers and pay for 1gb, rather than having 1 100mb transfer.  We'd also get none of the performance improvements.

My experience is that trusting the S3 docs is not always possible.  Just because they say that it's safe, doesn't mean I'll believe it until we have a test to prove that it is.